### PR TITLE
[WebUI] Add prompt format and stopping words for Qwen

### PR DESCRIPTION
**Pull request by sgwhat**
## Description

This PR adds support for prompt format and stopping words to the Qwen model, enabling correct output on the webUI. It also introduces the ability to specify stopping words for other models through the `custom_stopping_words` option.

### 1. Why the change?

Qwen model requires a prompt format to generate output correctly, and it also needs stopping words to halt the output when generation is complete.

### 2. User API changes
N/A

### 3. Summary of the change 
- Add prompt format for Qwen
- Add stopping words for Qwen
- Add `StopWordsCriteria` class for users to set stopping words manually through `custom_stopping_words` option.

### 4. How to test?
- [x] Application test

### CODE REVIEW

File: python/llm/example/Text-Generation-WebUI/modules/callbacks.py [hunk 1]
```diff
@@ -41,6 +41,20 @@ def __call__(self, input_ids: torch.LongTensor, _scores: torch.FloatTensor) -> b
...
+    def __init__(self, input_length, stop_words, tokenizer):
+        self.input_length = input_length
+        self.stop_words = stop_words
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        texts =  [ self.tokenizer.decode(ids[self.input_length:]) for ids in input_ids ]
+        dones = [ any(stop_word in text for stop_word in self.stop_words) for text in texts ]
+        return all(dones)
```
Comment (hkvision) [inline python/llm/example/Text-Generation-WebUI/modules/callbacks.py:55]: Do we need to check all new tokens or only the last token?

Comment (sgwhat) [reply python/llm/example/Text-Generation-WebUI/modules/callbacks.py:55]: all new tokens I thinks so.

Comment (hkvision) [reply python/llm/example/Text-Generation-WebUI/modules/callbacks.py:55]: will this impact the performance? We only need to check the last token each time?

File: python/llm/example/Text-Generation-WebUI/modules/callbacks.py [hunk 2]
```diff
@@ -41,6 +41,20 @@ def __call__(self, input_ids: torch.LongTensor, _scores: torch.FloatTensor) -> b
         return shared.stop_everything
 
 
+class StopWordsCriteria(transformers.StoppingCriteria):
+    """Custom `StoppingCriteria` which checks if all generated functions in the batch are completed."""
+    def __init__(self, input_length, stop_words, tokenizer):
+        self.input_length = input_length
```
Comment (hkvision) [inline python/llm/example/Text-Generation-WebUI/modules/callbacks.py:47]: this arg is no longer used?

File: python/llm/example/Text-Generation-WebUI/modules/callbacks.py [hunk 3]
```diff
@@ -41,6 +41,20 @@ def __call__(self, input_ids: torch.LongTensor, _scores: torch.FloatTensor) -> b
...
+        self.stop_words = stop_words
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        texts =  [self.tokenizer.decode(ids[-1]) for ids in input_ids]
```
Comment (hkvision) [inline python/llm/example/Text-Generation-WebUI/modules/callbacks.py:53]: Can we just check input_ids[-1][-1]?

File: python/llm/example/Text-Generation-WebUI/modules/callbacks.py [hunk 4]
```diff
@@ -41,6 +41,20 @@ def __call__(self, input_ids: torch.LongTensor, _scores: torch.FloatTensor) -> b
...
+class StopWordsCriteria(transformers.StoppingCriteria):
+    """Custom `StoppingCriteria` which checks if all generated functions in the batch are completed."""
+    def __init__(self, input_length, stop_words, tokenizer):
+        self.input_length = input_length
+        self.stop_words = stop_words
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        texts =  [self.tokenizer.decode(input_ids[-1][-1]) for ids in input_ids]
```
Comment (hkvision) [inline python/llm/example/Text-Generation-WebUI/modules/callbacks.py:53]: no need for loop? texts -> text is a single word?

File: python/llm/example/Text-Generation-WebUI/modules/callbacks.py [hunk 5]
```diff
@@ -41,6 +41,19 @@ def __call__(self, input_ids: torch.LongTensor, _scores: torch.FloatTensor) -> b
...
+class StopWordsCriteria(transformers.StoppingCriteria):
+    """Custom `StoppingCriteria` which checks if all generated functions in the batch are completed."""
+    def __init__(self, stop_words, tokenizer):
+        self.stop_words = stop_words
+        self.tokenizer = tokenizer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        """Returns true if all generated sequences contain any of the end-of-function strings."""
+        texts =  [self.tokenizer.decode(input_ids[-1][-1])]
+        dones = [any(stop_word in texts for stop_word in self.stop_words)]
```
Comment (hkvision) [inline python/llm/example/Text-Generation-WebUI/modules/callbacks.py:53]: text = self.tokenizer.decode(input_ids[-1][-1])
return text in self.stop_words
??

**Review by hkvision — APPROVED** [review]

**Pull request merged**